flink-clients/src/main/java/org/apache/flink/client/deployment/application/EmbeddedJobClient.java

```diff
@@ -126,7 +126,8 @@ public JobID getJobID() {
 					try {
 						return jobResult.toJobExecutionResult(classLoader);
 					} catch (Throwable t) {
-						throw new CompletionException(new Exception("Job " + jobId + " failed", t));
+						throw new CompletionException(
+								UnsuccessfulTerminationException.fromJobResult(jobResult, classLoader));
```

Comment:
Should this become part of the `JobClient.getJobExecutionResult` contract? If yes, then it needs to be documented. Moreover, one could also think about adding a unit test for this behaviour.

Is it guaranteed that the `ApplicationDispatcherBootstrap` will only work with an `EmbeddedJobClient`?

Reply:
For the first part, for now this is not the contract. Other implementations of the `JobClient` do not throw this exception and I did not like to change it in case people have exception handlers based on the current behaviour. Currently the `ClusterClientJobClientAdapter` wraps the exception from the `JobResult.toJobExecutionResult()` into a `ProgramInvocationException` and the `MiniClusterJobClient` simply wraps it in a `CompletionException`. So I would propose making this uniform and part of the contract in another issue.

That said, I believe that an exception with the job status like the `UnsuccessfulExecutionException` is a good candidate for the common behaviour.

For the second part, yes the `ApplicationMode` only works with an `EmbeddedJobClient`.

Reply:
I can open a separate issue and discuss it there.

Reply:
This is the issue https://issues.apache.org/jira/browse/FLINK-20020